### PR TITLE
Track and Update Completed Quests

### PR DESCRIPTION
**Pull request by ziyadbagh**
### **PR Description: Completed Quests Tracking and Updates**

This pull request introduces the following updates and features to enhance the functionality for tracking and managing completed quests:

---

#### **1. Added Quest Completion Methods in Repository and Interface**
- **Interface Updates:**
  - Added `getCompletedQuests` and `markQuestAsCompleted` functions to the `UserRepository` interface.

- **Repository Implementation:**
  - Implemented `getCompletedQuests` and `markQuestAsCompleted` in both the `MockUserRepository` and `FirestoreUserRepository` to handle tracking of completed quests.

---

#### **2. Enhanced UserViewModel for Quest Completion Tracking**
- **New Functions:**
  - Added functions to fetch and update completed quests in the `UserViewModel`.
  - Ensures LiveData observes updates to completed quests, enabling dynamic UI changes.

---

#### **3. Updated QuestScreen for Completed Quests**
- **UI Integration:**
  - The `QuestScreen` now fetches and displays completed quests dynamically based on data from the `UserViewModel`.
  - Added visual indicators for completed quests and updated the logic for unlocking new quests.

- **Progress Updates:**
  - When a quest is marked as complete, the screen reflects the changes immediately by updating both the quest status and the user stats.

---

#### **4. Stats Integration**
- **Stat Updates:**
  - Integrated completed quest tracking with the `StatsViewModel`.
  - Stats are updated in real-time as quests are completed, contributing to progress visualization.

---

#### **5. Comprehensive Testing**
- **Repository Tests:**
  - Added tests for `getCompletedQuests` and `markQuestAsCompleted` in the repository implementations to ensure correct functionality.

- **ViewModel Tests:**
  - Verified that the `UserViewModel` correctly handles fetching and updating completed quests.
  - Ensured the `StatsViewModel` reflects the updated stats after quest completion.

- **Screen Tests:**
  - Updated older tests for `QuestScreen`.
  
  Here is a demo:
  

https://github.com/user-attachments/assets/67553d68-0355-4bb6-82a2-60caa127024f

### CODE REVIEW

**Review by nada-ezziti — CHANGES_REQUESTED** [review]
This pull request is an excellent enhancement to the quest management functionality in the application, ensuring seamless tracking and updates for completed quests. It demonstrates a well-thought-out approach, integrating data persistence, dynamic UI updates, and thorough testing.

Adding getCompletedQuests and markQuestAsCompleted methods improves modularity and aligns with clean architecture principles.

The tests for repository methods are well-structured, simulating both success and failure scenarios.

After thoroughly reviewing the entire PR, everything looks good to me. Once the SonarCloud issue is resolved and the hardcoded string is removed, I believe this PR is ready to be merged.

File: app/src/main/java/com/github/se/signify/ui/screens/home/QuestScreen.kt [hunk 1]
```diff
@@ -195,14 +224,22 @@ fun QuestDescriptionDialog(
...
-                    onClick = { isFingerspellVisible = true },
-                    colors =
-                        ButtonDefaults.buttonColors(
-                            containerColor = MaterialTheme.colorScheme.secondary,
-                            contentColor = MaterialTheme.colorScheme.onSecondary)) {
-                      Text(stringResource(R.string.try_fingerspell_button_text))
-                    }
+                if (isCompleted) {
+                  Text(
+                      text = "Completed",
```
Comment (nada-ezziti) [inline app/src/main/java/com/github/se/signify/ui/screens/home/QuestScreen.kt:229]: You should move this hardcoded value to the file strings.xml

**Review by nada-ezziti — DISMISSED** [review]

File: app/src/main/java/com/github/se/signify/ui/screens/home/QuestScreen.kt (hunk 1 above)
Comment (ziyadbagh) [reply app/src/main/java/com/github/se/signify/ui/screens/home/QuestScreen.kt:229]: Good catch!

**Review by nada-ezziti — APPROVED** [review]
Everything seems good, we can merge this PR, good job!

**Pull request merged**